src/xcode/ENA/ENA/Source/Services/Risk/Provider/RiskProvider.swift

```diff
@@ -107,11 +108,28 @@ final class RiskProvider: RiskProviding {
 
 			// Keep downloading key packages and trace warning packages for plausible deniability
 
-			downloadKeyPackages()
-
-			appConfigurationProvider.appConfiguration().sink { [weak self] appConfiguration in
-				self?.downloadTraceWarningPackages(with: appConfiguration, completion: { _ in })
-			}.store(in: &subscriptions)
+			downloadKeyPackages { [weak self] _ in
+				guard let self = self else {
+					return
+				}
+				self.appConfigurationProvider.appConfiguration().sink { [weak self] appConfiguration in
+					self?.downloadTraceWarningPackages(with: appConfiguration, completion: { [weak self] result in
+						guard let self = self else {
+							return
+						}
+						switch result {
+						case .success:
+							if let risk = self.previousRiskIfExistingAndNotExpired(userInitiated: userInitiated) {
+								self.successOnTargetQueue(risk: risk)
```

Comment:
Do I understand in correctly?
After the download of key packages and warning packages, success is called with the previously calculated risk.
And then the calculation of the risk is triggered, which the also may return success with the new calculated risk.

To avoid race conditions, why is the logic not simply calling "_requestRiskLevel" inside of download completion?

Reply:
I missed that its in a guard with return. Cleared.